Categorized latejoin menu and made it into VueUI, improved crew manifest logic and made Verb manifest same as VueUI program, Job fixes

Amunak opened the pull request:
- Latejoin Menu and Crew Manifest are both VueUI and they are prettier and similar-er.
- All departmental jobs now have a department assigned (intern jobs and forensic technician). This was an oversight I believe; it slightly changed the antagonist report generated at round start. Now the fix is especially needed for the latejoin menu sorting to work (they'd be in the miscelleanous section otherwise).
- Department names are now defines, crew manifest logic has been simplified to accommodate this. The 'Heads' section is now called 'Command'.
- All crew manifests now open as the VueUI; this makes them consistent and pretty everywhere.
- Manifest now asks SSJobs for the job by rank instead of trying to guess it itself based on rank names.
- The crew manifest program can now be printed from its UI (at consoles with printers).
- ID modification program no longer shows or prints the crew manifest.
- Moved the SCSS department color "defines" to a separate file, made them globally available as a map and shortcuts for text and background color.
- Added SCSS utilities for simple styling with just classes, inspired by Bootstrap. This means we can use classes like `text-center`, `fs-large` (large font size), `mt-2` (margin-top with size of spacer 2), ... If this goes through I'll probably look around the rest of the UIs in an attempt to make them more consistent and use these classes.
- Jobs now store departments as a list so Command is no longer magic. Flags can be assigned to each department (like 'this is a supervisor/head of this department).'
- Leaving the round now actually decreases the job count for a given job, freeing up the slot for others. This also fixes wrong people counters in latejoin menu. As a result AIs now use different logic to check if they can spawn/latejoin and their numbers are also correct.
- Merchants no longer have a department. This should affect nothing.
- Occupations settings in character creation now use the same department colors and the buttons are more readable.

### Screenshots
Note that some of the screenshots are a bit outdated now; QM does not show up in Command anymore.

#### Latejoin menu
![image](https://user-images.githubusercontent.com/781546/98733812-6409c200-23a1-11eb-9573-3b6e99e26b17.png)

#### Crew Manifest
![2020-11-08_22-35-32](https://user-images.githubusercontent.com/781546/98484923-14958b80-2213-11eb-8503-162b5ebe74fb.png)

#### Printed manifest
![dreamseeker_2020-11-08_00-27-16](https://user-images.githubusercontent.com/781546/98469802-109d4580-21e2-11eb-84a5-41c46d245229.png)

#### Occupations setup
![image](https://user-images.githubusercontent.com/781546/98732856-f8732500-239f-11eb-848f-14dadf878499.png)

Amunak commented:
!review

mikomyazaki commented:
Did you consider using departmental colors?

mikomyazaki commented on code/controllers/subsystems/records.dm:
```diff
@@ -239,56 +237,46 @@
+
+		 // add them to their department
+		manifest[dept][++manifest[dept].len] = list("name" = name, "rank" = rank, "active" = isactive, "head" = depthead)
+		if(dept == DEPARTMENT_COMMAND) // they are a captain, put them on top of command
```
1. Isn't this also true if you're a department head?

2. It feels kind of weird to have `DEPARTMENT_COMMAND` mean two different things, since it is used as a key in `positions_by_department` and as a var on `/datum/job` - It doesn't mean the same thing for the Captain as it does for a Department Head. 

Since this line is supposed to uniquely identify the Captain job, it should be something like `rank == "Captain"` or some map-defined head-of-heads var that would be Captain.

Amunak replied:
1. It is not, department heads have the department they are a head of as ... their department. Well technically it _is_ true but thanks to the code above (and the fact that Command is first) they get assigned Command first and then it's immediately overwritten with their actual department.
2. It _is_ weird but command is just special that way. It's not a real department since the only person actually assigned to it is the captain; everything else is "copies". I don't see a way out of this short of completely reworking how we think about Command and allowing people to _actually_ be in multiple departments at once while being a head in one and not the other.
3. I specifically left it this way to (1) not have to hardcode the Captain rank here and (2) if it ever happens that we have some proper member in there _other than captain_ (like a captain's / command secretary or something) it would get assigned in there properly.

Overall this is something I would prefer not to change; I had to touch it mostly because I added the constants and had to get rid of that weird "map".

Amunak replied:
Also fixed by asking SSJobs for the job.

mikomyazaki commented on code/controllers/subsystems/records.dm:
```diff
@@ -239,56 +237,46 @@
+			if(department == DEPARTMENT_COMMAND)
+				depthead = TRUE
+				continue
+			// because Command is first we have guarantee that the depthead var has been assigned and we have found the actual dept so we can safely break
```
I don't really like that this all relies on the ordering of `positions_by_department`. 

We have this 'department_flag` thing, too, that we're ignoring. 

Could we define heads as being e.g. `flag = CIVILIAN | HEAD` and check that var instead?

I don't know why we have both 'department' and 'department_flag'

Amunak replied:
> I don't really like that this all relies on the ordering of positions_by_department.

Me neither, but as stated above Command is just ... special, and I'm not sure it would be easy to deal with that.

---

I'll take a look at the flags. But I think we need the constants regardless, for output procs (like the manifest) if nothing else.

Amunak replied:
Fixed by asking SSJobs for the job instead of this nightmare.

mikomyazaki commented on code/controllers/subsystems/records.dm:
```diff
@@ -297,12 +285,11 @@
 				selected_module = capitalize_first_letters(R.module.name)
-			manifest["bot"][++manifest["bot"].len] = list("name" = sanitize(R.name), "rank" = selected_module, "active" = "Online")
+			manifest[dept][++manifest[dept].len] = list("name" = sanitize(R.name), "rank" = selected_module, "active" = "Online", "head" = FALSE)
 		if(istype(S, /mob/living/silicon/ai))
```
```suggestion
		else if(istype(S, /mob/living/silicon/ai))
```

mikomyazaki commented on code/game/jobs/job/captain.dm:
```diff
@@ -3,7 +3,7 @@ var/datum/announcement/minor/captain_announcement = new(do_newscast = 1)
 	flag = CAPTAIN
-	department = "Command"
+	department = DEPARTMENT_COMMAND
 	head_position = 1
```
So we have two completely separate ways of telling if someone is a head?

1. If they're in `positions_by_department[DEPARTMENT_COMMAND]`
2. If they have `head_position = TRUE`

One needs to go - Or both do & turn things into flags like I suggested above.

Amunak replied:
Someone is a head for the department set in `department` if they have `head_position = true`. Technically you can _tell_ that someone is a head by looking into DEPARTMENT_COMMAND, but that will not necessarily be true in the future, and if that check is used somewhere it should be gone, you're right.

I'll fix the manifest code that does this.

Amunak replied:
Manifest code rewritten. I could just remove `positions_by_department` since it's not needed anymore.

mikomyazaki commented on code/game/jobs/jobs.dm:
```diff
@@ -132,6 +132,18 @@ var/list/nonhuman_positions = list(
 	"Merchant"
 )
 
+var/list/positions_by_department = list(
```
```suggestion
var/list/list/positions_by_department = list(
```

mikomyazaki commented on code/modules/mob/abstract/new_player/new_player.dm:
```diff
@@ -323,21 +327,37 @@ INITIALIZE_IMMEDIATE(/mob/abstract/new_player)
 	for(var/datum/job/job in SSjobs.occupations)
 		if(job && IsJobAvailable(job.title))
+			var/department = job.department
+			if(!department || !(department in jobs_by_department)) // no department set or it's something weird
```
```suggestion
			if(!(department in jobs_by_department)) // no department set or it's something weird
```

mikomyazaki commented on code/modules/mob/abstract/new_player/new_player.dm:
```diff
@@ -323,21 +327,37 @@ INITIALIZE_IMMEDIATE(/mob/abstract/new_player)
+			var/department = job.department
+			if(!department || !(department in jobs_by_department)) // no department set or it's something weird
+				department = DEPARTMENT_MISCELLANEOUS
+			if(job.head_position) // make sure heads are first
```
See above about department flags. Wouldn't it be nicer to have

```suggestion
#define JOB_IS_HEAD (job.department_flag & HEAD)
...
			if(JOB_IS_HEAD) // make sure heads are first
```

Or something?

Amunak replied:
I'm not sure if this is the best approach - mixing different types of values (jobs and "is head") together. Like, if we wanted to assign Heads "properly" to both their department _and_ command this would cause issues because they are a head in their department but not in Command.

But we definitely need to have just one place to check for these (as well as for departments) otherwise there will be some inconsistency somewhere. I just don't think I should fix that in this PR, it's already ginormous.

mikomyazaki commented on code/modules/mob/abstract/new_player/new_player.dm:
```diff
@@ -323,21 +327,37 @@ INITIALIZE_IMMEDIATE(/mob/abstract/new_player)
+		for(var/datum/job/job in jobs_by_department[department])
 			var/active = 0
 			// Only players with the job assigned and AFK for less than 10 minutes count as active
 			for(var/mob/M in player_list) //Added isliving check here, so it won't check ghosts and qualify them as active
 				if(isliving(M) && M.mind && M.client && M.mind.assigned_role == job.title && M.client.inactivity <= 10 MINUTES)
```
```suggestion
			for(var/mob/living/M in player_list)
				if(M.mind && M.client && M.mind.assigned_role == job.title && M.client.inactivity <= 10 MINUTES)
```

Obviously this isn't your doing, but it bothers me :P

Amunak commented:
> Did you consider using departmental colors?

I did but that would mean converting the latejoin menu to VueUI because that's where we have the colors.

mikomyazaki commented on code/controllers/subsystems/records.dm:
```diff
@@ -179,54 +179,26 @@
+	. = ..()
+	data = . || data || list()
+
+	VUEUI_SET_CHECK(data["manifest"], SSrecords.get_manifest_list(), ., data)
```
`VUEUI_SET_CHECK` does not work for lists, it will always set. There is a separate helper for lists.

Alternatively just don't use a SET_CHECK because they are largely unnecessary.

mikomyazaki commented on code/modules/modular_computers/file_system/programs/generic/manifest.dm:
```diff
@@ -30,16 +30,24 @@
 		. = data
 
+	VUEUI_SET_CHECK_IFNOTSET(data["allow_printing"], !!(computer?.nano_printer), ., data)
 	VUEUI_SET_CHECK(data["manifest"], SSrecords.get_manifest_list(), ., data)
```
Same here.

mikomyazaki commented on code/modules/modular_computers/file_system/programs/generic/manifest.dm:
```diff
@@ -30,16 +30,24 @@
 /datum/computer_file/program/manifest/Topic(href, href_list)
 	. = ..()
 
+	if(href_list["action"] == "print" && can_run(usr, 1) && !!(computer?.nano_printer))
```
```suggestion
	if(href_list["action"] == "print" && can_run(usr, 1) && (computer?.nano_printer))
```

mikomyazaki commented on code/modules/modular_computers/file_system/programs/generic/manifest.dm:
```diff
@@ -30,16 +30,24 @@
 
+	if(href_list["action"] == "print" && can_run(usr, 1) && !!(computer?.nano_printer))
+		if(!computer.nano_printer.print_text(SSrecords.get_manifest_text(), text("crew manifest ([])", worldtime2text())))
+			to_chat(usr, SPAN_WARNING(">Hardware error: Printer was unable to print the file. It might be out of paper."))
```
Is that `>` a result of you copy-pasting SPAN_WARNINGs and messing up with a span?

Amunak replied:
As with most of the other issues it's caused by just copying the [original code](https://github.com/Aurorastation/Aurora.3/pull/10470/files#diff-f184093de1da6533b4c920bf39996da0a13581d535cc85516bdb0e0819ae50d6L153) and not thinking about it / seeing it. x)

mikomyazaki commented on code/datums/late_choices.dm:
```diff
@@ -0,0 +1,90 @@
+	var/jobs_available = 0
+	var/list/list/datum/job/jobs_by_department = DEPARTMENTS_LIST_INIT
+	for(var/datum/job/job in SSjobs.occupations)
+		if(job && NP.IsJobAvailable(job.title))
```
```suggestion
		if(NP.IsJobAvailable(job.title))
```

Job should always exist here due to the filter in the for loop.

Although... SSjobs.occupations should always contain only /datum/job so...

```suggestion
for(var/J in SSjobs.occupations)
		var/datum/job/job = J
		if(job && NP.IsJobAvailable(job.title))
```

Maybe that is more appropriate?

Amunak replied:
I think I'll go with the first variant, it's more ... "familiar"? You don't have to rely on the type of the list which is in a different file somewhere.

mikomyazaki commented on code/controllers/subsystems/records.dm:
```diff
@@ -181,54 +179,27 @@
+		return STATUS_INTERACTIVE
+	if(isobserver(user))
+		return STATUS_INTERACTIVE
+	if(issilicon(user)) // silicons have the show manifest verb
+		return STATUS_INTERACTIVE
```
Should this, then, either call their verb or display a message telling them to use their verb? Not sure.

Amunak replied:
Oh no, this is for when this code is called _by_ their verb. It's there just to remind people that silicons do have this despite also having a computer inside.

mikomyazaki commented on vueui/src/components/view/late-choices.vue:
```diff
@@ -0,0 +1,100 @@
+      <p>
+        Welcome, <strong>{{ character_name }}</strong>.<br>
+        Round duration: <strong>{{ round_duration }}</strong><br>
+        Alert level: <strong :style="{color: alertLevelColor }">{{ alert_level }}</strong><br>
```
I see in the preview pictures you provided that this isn't green when it displays green?

Amunak replied:
Indeed, you can see the `alertLevelColor` function in the bottom. My reasoning was that we want to highlight only "unusual" states, so green is just regular text color.

Amunak commented:
This should be (I hope) the final version (barring issues). QM no longer shows up in Command, fixed some of the Jobs BS.

Additionally I used the department colors for character creation's occupation setup and improved visibility of the buttons there:

![image](https://user-images.githubusercontent.com/781546/98732856-f8732500-239f-11eb-848f-14dadf878499.png)

NonQueueingMatt commented on code/modules/mob/abstract/new_player/new_player.dm:
```diff
@@ -5,6 +5,7 @@
 	var/spawning = 0//Referenced when you want to delete the new_player later on in the code.
 	var/totalPlayers = 0		 //Player counts for the Lobby tab
 	var/totalPlayersReady = 0
+	var/datum/late_choices/late_choices_ui = null
```
clean this up in destroy

Amunak replied:
Do you mean just setting the reference to null? I didn't bother since the UI gets closed automatically - transferring the player to a new mob closes all the UIs automatically, and that cleans it up. But I guess removing the reference there wouldn't hurt.

NonQueueingMatt commented on code/modules/mob/abstract/new_player/new_player.dm:
```diff
@@ -295,50 +296,12 @@ INITIALIZE_IMMEDIATE(/mob/abstract/new_player)
+	else
+		// This creates the latechoices UI.
+		// It is automatically created and then destroyed and unassigned when closed.
+		late_choices_ui = new(src)
```
Since these are per-player is it not possible to cache the gFI icon on the datum?

Amunak replied:
There's no difference, the datum disappears as soon as the dialogue closes.

And the reason why I made it refresh on (re-)opening is so that you can change the character in character loadout and see the change.

_Maybe_ it could be done so that the UI refreshes all the time (to see character, alert level and whatnot changes immediately) _but_ cache the icon and update it only when the character actually changes. But that would still update it as much as now; we'd just get the benefits of the constantly updating UI.

Amunak commented on code/datums/late_choices.dm:
```diff
@@ -0,0 +1,92 @@
+/datum/late_choices/proc/ui_open(mob/user)
+	ui = new(user, src, "late-choices", 330, 720, "Late-Join Choices")
+	ui.header = "minimal"
+	var/mob/mannequin = user.client.prefs.update_mannequin()
+	ui.add_asset("character", getFlatIcon(mannequin, SOUTH))
```
That's what I wanted to do originally, but the character prefs use a screen object and /image to render it, and `ui.add_asset` can push only /icons.

Amunak commented:
The UI datum is now "persistent" while the new player mob is alive. This means that the mob icon gets cached and generated only once, _unless_ a change is generated by the character setup. Moreover the icon is regenerated only when actually needed (seen) by the player and not every time it changes in the setup UI.

I also made the UI refresh constantly so that people can see slots (dis)appearing, alert level changes, etc. Hope that addresses the issues.

Oh and the UI should now delete itself properly, though both this and the constant changing seems to be broken when the UI is closed and re-opened due to some bug in the VueUI handling (_I hope_).
